cql3/expr/expression.hh

```diff
@@ -136,11 +136,10 @@ using value_set = std::variant<value_list, nonwrapping_range<bytes>>;
 extern value_set possible_lhs_values(const column_definition*, const expression&, const query_options&);
 
 /// Turns value_set into a range, unless it's a multi-valued list (in which case this throws).
-extern nonwrapping_range<bytes> to_range(const value_set&);
+extern nonwrapping_range<managed_bytes> to_range(const value_set&);
 
 /// A range of all X such that X op val.
-template<typename T>
-nonwrapping_range<T> to_range(oper_t op, const T& val);
+nonwrapping_range<clustering_key_prefix> to_range(oper_t op, const clustering_key_prefix& val);
```

Comment:
This isn't quite what I meant.  Why couldn't we keep the declaration templated, just with a `T&&` parameter?  And the specialized `to_range<clustering_key_prefix>` declaration in expressions.cc wouldn't need a body.  A templated `to_range` is more widely useful than one that only takes a `clustering_key_prefix` parameter.

Reply:
Template declarations are tricky to maintain, because you have to manually keep track of all uses and explicitly instantiate them all. If you don't, linking may succeed by accident (because an implicit instantiation happened to be generated), and then fail in a different build mode or a few patches later.
With forwarding references they are double tricky, because now `T` can also be a reference type. In other words, we would have to explicitly instantiate not only `to_range<clustering_key_prefix>`, but also `to_range<const clustering_key_prefix&>`, `to_range<clustering_key_prefix&>` and `to_range<clustering_key_prefix&&>`, depending on which of them are actually used. I think it's more reasonable to declare concrete versions in the header and implement them using the template.
Alternatively we could just move the implementation to the header.

Reply:
> Template declarations are tricky to maintain, because you have to manually keep track of all uses and explicitly instantiate them all.

Yes, that's precisely what I was explicitly [asked](https://github.com/scylladb/scylla/pull/8227#discussion_r589007986) to do with `to_range`.  Perhaps @avikivity should chime in.

> Alternatively we could just move the implementation to the header.

That's how it started, but I was advised against it. :shrug:

Reply:
I don't have the context here.

Reply:
Context: this `to_range` function translates a binary_operator into an interval.  I templatized it so it can be used on multiple types.  You asked me to move the definition into the .cc file.  Now @michoecho wants to either detemplatize it or move the definition back into the header.

Reply:
Well, de-templating is the best (as long as it's not replaced with manual duplication).

Second best is templates-but-not-in-the-header (as long as it doesn't involve long lists of explicit instantiations).

Third best is templates in the header.

Reply:
> Well, de-templating is the best (as long as it's not replaced with manual duplication).

But it will be.  The very next thing I'll need is `to_range` for partition ranges.

Reply:
To clarify my doubt: why would `foo(T&&)` ever instantiate into anything other than `foo(int&&)` for any invocation of foo on an integer reference?

Reply:
That's because `T&&` is not an rvalue reference, but a forwarding reference.

Example:
```
int a;
const int b;

template <typename T> f(const T& x) {}
f(a); // T is deduced to int
f(b); // T is deduced to int
f(0); // T is deduced to int

// However...

template <typename T> g(T&& x) {}
g(a); // T is deduced to int&
g(b); // T is deduced to const int&
g(0); // T is deduced to int
```

See https://en.cppreference.com/w/cpp/utility/forward.

Reply:
@michoecho since you got lots of work to do elsewhere in this PR, maybe it's best to leave this as it is now (ie, detemplatized).  It tops Avi's goodness scale when considered individually.  If/when the future arrives where I need multiple copies, I'll templatize it again.

Reply:
Okay. But why would you need a public template, anyway? If you only need 2-3 versions of this function (I thought that was the case?), then isn't it the best two just write 2-3 declarations in the header, one template in the `.cc`, and implement those public declarations using the template, as I did in this patch? I don't see the downsides. There is some duplication, but we are only duplicating the declarations (and the one-liner bodies in .cc), not the logic.

Reply:
> isn't it the best to just write 2-3 declarations in the header, one template in the `.cc`, and implement those public declarations using the template

No, that's second best. ;) It's strictly more typing than 1 header declaration and 2-3 bodyless instantiation declarations in the .cc file.  Both initially and upon every future change of function name or signature.

Reply:
OTOH, it's easier to forget to instantiate something (especially with forward references, apparently) and cause a problem that only becomes visible two months later for someone else...